dom: Implement `WritableStream`

gterzian opened the pull request:
<!-- Please describe your changes on the following line: -->

Implement writable stream https://streams.spec.whatwg.org/#writablestream

Part of https://github.com/servo/servo/issues/34676

---
<!-- Thank you for contributing to Servo! Please replace each `[ ]` by `[X]` when the step is complete, and replace `___` with appropriate data: -->
- [ ] `./mach build -d` does not report any errors
- [ ] `./mach test-tidy` does not report any errors
- [ ] These changes fix #___ (GitHub issue number if applicable)

<!-- Either: -->
- [ ] There are tests for these changes OR
- [ ] These changes do not require tests because ___

<!-- Also, please make sure that "Allow edits from maintainers" checkbox is checked, so that we can help you if you get stuck somewhere along the way.-->

<!-- Pull requests that do not address these steps are welcome, but they will require additional verification as part of the review process. -->

gterzian committed:
add basic interface for writable stream

Signed-off-by: gterzian <2792687+gterzian@users.noreply.github.com>

gterzian committed:
add stubs for pipeTo and pipeThrough methods

Signed-off-by: gterzian <2792687+gterzian@users.noreply.github.com>

gterzian committed:
add stubs for writable stream defautl writer

Signed-off-by: gterzian <2792687+gterzian@users.noreply.github.com>

gterzian committed:
add stubs for writable stream controller

Signed-off-by: gterzian <2792687+gterzian@users.noreply.github.com>

gterzian committed:
add underlying source dict

Signed-off-by: gterzian <2792687+gterzian@users.noreply.github.com>

gterzian committed:
add underlying source dict

Signed-off-by: gterzian <2792687+gterzian@users.noreply.github.com>

gterzian committed:
implement constructor

Signed-off-by: gterzian <2792687+gterzian@users.noreply.github.com>

gterzian committed:
implement init writable stream

Signed-off-by: gterzian <2792687+gterzian@users.noreply.github.com>

gterzian committed:
impl setup default controller

Signed-off-by: gterzian <2792687+gterzian@users.noreply.github.com>

gterzian committed:
implement controller setup

Signed-off-by: gterzian <2792687+gterzian@users.noreply.github.com>

gterzian committed:
implement controller advance queue if neededd

Signed-off-by: gterzian <2792687+gterzian@users.noreply.github.com>

gterzian committed:
implement stream finish erroring

Signed-off-by: gterzian <2792687+gterzian@users.noreply.github.com>

gterzian committed:
implement stream reject close and closed promise if needed

Signed-off-by: gterzian <2792687+gterzian@users.noreply.github.com>

gterzian committed:
finish implementation of stream finish erroring

Signed-off-by: gterzian <2792687+gterzian@users.noreply.github.com>

gterzian committed:
call into controller setup from stream constructor

Signed-off-by: gterzian <2792687+gterzian@users.noreply.github.com>

gterzian committed:
implement stream mark first write request in flight

Signed-off-by: gterzian <2792687+gterzian@users.noreply.github.com>

gterzian committed:
implement controller process write

Signed-off-by: gterzian <2792687+gterzian@users.noreply.github.com>

gterzian committed:
call into advance queue if needed at various points

Signed-off-by: gterzian <2792687+gterzian@users.noreply.github.com>

gterzian committed:
implement stream deal with rejection, use from_safe_context

Signed-off-by: gterzian <2792687+gterzian@users.noreply.github.com>

gterzian committed:
implement controller clear algorithms

Signed-off-by: gterzian <2792687+gterzian@users.noreply.github.com>

gterzian committed:
remove unused todo

Signed-off-by: gterzian <2792687+gterzian@users.noreply.github.com>

gterzian committed:
implement stream start erroring

Signed-off-by: gterzian <2792687+gterzian@users.noreply.github.com>

gterzian committed:
finish writer ensure ready promise rejected

Signed-off-by: gterzian <2792687+gterzian@users.noreply.github.com>

gterzian committed:
implement stream finish in flight write request

Signed-off-by: gterzian <2792687+gterzian@users.noreply.github.com>

gterzian committed:
impl controller process close

Uses ai

Signed-off-by: gterzian <2792687+gterzian@users.noreply.github.com>

gterzian committed:
implement write constructor and setup

Signed-off-by: gterzian <2792687+gterzian@users.noreply.github.com>

gterzian committed:
implement controller error

Signed-off-by: gterzian <2792687+gterzian@users.noreply.github.com>

gterzian committed:
remove unnecessary unsafe code

Signed-off-by: gterzian <2792687+gterzian@users.noreply.github.com>

gterzian committed:
finish implementing process write

Signed-off-by: gterzian <2792687+gterzian@users.noreply.github.com>

gterzian committed:
implement close sentinel

Signed-off-by: gterzian <2792687+gterzian@users.noreply.github.com>

gterzian committed:
implement public locked

Signed-off-by: gterzian <2792687+gterzian@users.noreply.github.com>

gterzian committed:
implement stream abort

Signed-off-by: gterzian <2792687+gterzian@users.noreply.github.com>

gterzian committed:
implement stream close

Signed-off-by: gterzian <2792687+gterzian@users.noreply.github.com>

gterzian committed:
implement controller close

Signed-off-by: gterzian <2792687+gterzian@users.noreply.github.com>

gterzian committed:
fix use of crown

Signed-off-by: gterzian <2792687+gterzian@users.noreply.github.com>

gterzian committed:
remove unnecessary options around writer promises

Signed-off-by: gterzian <2792687+gterzian@users.noreply.github.com>

gterzian committed:
implement writer get desired size

Signed-off-by: gterzian <2792687+gterzian@users.noreply.github.com>

gterzian committed:
implement writer ready

Signed-off-by: gterzian <2792687+gterzian@users.noreply.github.com>

gterzian commented:
Note to self, apply rooting pattern of https://github.com/servo/servo/pull/34194/

gterzian committed:
implement writer abort

Signed-off-by: gterzian <2792687+gterzian@users.noreply.github.com>

gterzian committed:
implement writer close

Signed-off-by: gterzian <2792687+gterzian@users.noreply.github.com>

gterzian committed:
implement writer release lock

Signed-off-by: gterzian <2792687+gterzian@users.noreply.github.com>

gterzian committed:
implement writer public write

Signed-off-by: gterzian <2792687+gterzian@users.noreply.github.com>

gterzian committed:
implement private writer write

Uses ai

Signed-off-by: gterzian <2792687+gterzian@users.noreply.github.com>

gterzian committed:
implement writer release.

Uses ai

Signed-off-by: gterzian <2792687+gterzian@users.noreply.github.com>

gterzian committed:
finish controller process close

Signed-off-by: gterzian <2792687+gterzian@users.noreply.github.com>

gterzian committed:
root promise handlers

Signed-off-by: gterzian <2792687+gterzian@users.noreply.github.com>

gterzian committed:
handler errors in stream and writer constructor
finish implementation of stream finish in flight close

Signed-off-by: gterzian <2792687+gterzian@users.noreply.github.com>

gterzian committed:
fix warnings

Signed-off-by: gterzian <2792687+gterzian@users.noreply.github.com>

gterzian committed:
implement controller get chunk size

Signed-off-by: gterzian <2792687+gterzian@users.noreply.github.com>

gterzian committed:
tidy the webidls

Signed-off-by: gterzian <2792687+gterzian@users.noreply.github.com>

gterzian committed:
implement stream get writer

Signed-off-by: gterzian <2792687+gterzian@users.noreply.github.com>

gterzian committed:
fix assertion of stream state when advancing queue if needed

Signed-off-by: gterzian <2792687+gterzian@users.noreply.github.com>

gterzian committed:
add docs for value with size

Signed-off-by: gterzian <2792687+gterzian@users.noreply.github.com>

gterzian committed:
use reject_error in abort

Signed-off-by: gterzian <2792687+gterzian@users.noreply.github.com>

gterzian committed:
remove unnecessary allowances of unsafe code

Signed-off-by: gterzian <2792687+gterzian@users.noreply.github.com>

gterzian committed:
turn writable-streams test suite on

Signed-off-by: gterzian <2792687+gterzian@users.noreply.github.com>

gterzian committed:
update encodings test expectations

Signed-off-by: gterzian <2792687+gterzian@users.noreply.github.com>

gterzian committed:
properly check if type is set on sink in stream constructor

Signed-off-by: gterzian <2792687+gterzian@users.noreply.github.com>

gterzian committed:
fix double borrow in controller advance queue if needed

Signed-off-by: gterzian <2792687+gterzian@users.noreply.github.com>

gterzian committed:
make the queue aware of the close sentinel when dequeuing a value

Signed-off-by: gterzian <2792687+gterzian@users.noreply.github.com>

gterzian committed:
fix assertion of no backpressure in update backpressure

Signed-off-by: gterzian <2792687+gterzian@users.noreply.github.com>

gterzian committed:
also clear strategy size when clearing algorithms

Signed-off-by: gterzian <2792687+gterzian@users.noreply.github.com>

gterzian committed:
remove this object arg when calling into strategy size

Signed-off-by: gterzian <2792687+gterzian@users.noreply.github.com>

gterzian committed:
fix has operations marked in flight

Signed-off-by: gterzian <2792687+gterzian@users.noreply.github.com>

gterzian committed:
fix typo in has in flight write request

Signed-off-by: gterzian <2792687+gterzian@users.noreply.github.com>

gterzian committed:
turn error into no-op when aborting a stream, if the stream is closed or errored.

Signed-off-by: gterzian <2792687+gterzian@users.noreply.github.com>

gterzian committed:
fix error handling of calling into abort algorithm

Signed-off-by: gterzian <2792687+gterzian@users.noreply.github.com>

gterzian committed:
fix error handling of calling into close and write algorithms

Signed-off-by: gterzian <2792687+gterzian@users.noreply.github.com>

gterzian committed:
fix double borrow on queue
fix logic in update_backpressure
fix logic in get_desired_size
fix logic in writer setup

Signed-off-by: gterzian <2792687+gterzian@users.noreply.github.com>

servo-wpt-sync commented:
🤖 Opened new upstream WPT pull request (web-platform-tests/wpt#50571) with upstreamable changes.

servo-wpt-sync commented:
🤖 This change no longer contains upstreamable changes to WPT; closed existing upstream pull request (web-platform-tests/wpt#50571).

gterzian committed:
update test expectations for aborting suite

Signed-off-by: gterzian <2792687+gterzian@users.noreply.github.com>

gterzian committed:
fix controller get_backpressure

Signed-off-by: gterzian <2792687+gterzian@users.noreply.github.com>

gterzian committed:
fix clippy

Signed-off-by: gterzian <2792687+gterzian@users.noreply.github.com>

gterzian committed:
update test expectations to expect errors in tests using unsupported apis

Signed-off-by: gterzian <2792687+gterzian@users.noreply.github.com>

gterzian committed:
fix error handling of calling into start algo in controller setup

Signed-off-by: gterzian <2792687+gterzian@users.noreply.github.com>

servo-wpt-sync commented:
🤖 Opened new upstream WPT pull request (web-platform-tests/wpt#50601) with upstreamable changes.

gterzian committed:
update test expectations for test checking for undefined this in strategy size call

Signed-off-by: gterzian <2792687+gterzian@users.noreply.github.com>

gterzian committed:
update test expectation to timeout for response-stream-with-broken-then.any.worker

Signed-off-by: gterzian <2792687+gterzian@users.noreply.github.com>

gterzian committed:
update interfaces

Signed-off-by: gterzian <2792687+gterzian@users.noreply.github.com>

gterzian committed:
fix use of global() and error to_jsval

Signed-off-by: gterzian <2792687+gterzian@users.noreply.github.com>

gterzian committed:
fix use of crown for promise handlers

Signed-off-by: gterzian <2792687+gterzian@users.noreply.github.com>

gterzian committed:
remove fail expectation from worker interface objects test

Signed-off-by: gterzian <2792687+gterzian@users.noreply.github.com>

gterzian committed:
remove fail expectation for test expecting this to be undefined in callback

Signed-off-by: gterzian <2792687+gterzian@users.noreply.github.com>

gterzian committed:
fix documentation link for writablestream state

Signed-off-by: gterzian <2792687+gterzian@users.noreply.github.com>

gterzian committed:
refactor write_requests to use a vec deque

uses ai

Signed-off-by: gterzian <2792687+gterzian@users.noreply.github.com>

gterzian committed:
remove unnecessary doc

Signed-off-by: gterzian <2792687+gterzian@users.noreply.github.com>

gterzian committed:
refactor reject_close_and_closed_promise_if_needed to take a safe js context as argument

uses ai

Signed-off-by: gterzian <2792687+gterzian@users.noreply.github.com>

gterzian committed:
pass globals and contexts by ref where possible

uses ai

Signed-off-by: gterzian <2792687+gterzian@users.noreply.github.com>

gterzian committed:
fix doc link for controller

Signed-off-by: gterzian <2792687+gterzian@users.noreply.github.com>

gterzian committed:
remove unnecessary comment

Signed-off-by: gterzian <2792687+gterzian@users.noreply.github.com>

gterzian committed:
change update_backpressure to be a method of the writablestream

uses ai

Signed-off-by: gterzian <2792687+gterzian@users.noreply.github.com>

gterzian committed:
rename writer method that resolve closed and ready promise for clarity

uses ai

Signed-off-by: gterzian <2792687+gterzian@users.noreply.github.com>

gterzian commented:
@Taym95 @jdm This one is now ready for review.

Taym95 reviewed (commented):
Initial review, I will follow up later

gterzian commented on components/script/dom/writablestream.rs:
```diff
@@ -0,0 +1,949 @@
+            let promise = controller.abort_steps(cx, global, reason.handle(), can_gc);
+
+            // Upon fulfillment of promise,
+            rooted!(in(*cx) let mut fulfillment_handler = Some(AbortAlgorithmFulfillmentHandler {
```
yes, it's how this works following @jdm introduction of it in https://github.com/servo/servo/pull/34194

gterzian commented on components/script/dom/writablestreamdefaultcontroller.rs:
```diff
@@ -0,0 +1,834 @@
+        // Set stream.[[controller]] to controller.
+        stream.set_default_controller(self);
+
+        // Perform ! ResetQueue(controller).
```
All those steps are on done earlier, as noted below with `Note: above steps are done in `new_inherited`.`

gterzian commented on tests/wpt/meta/streams/writable-streams/aborting.any.js.ini:
```diff
@@ -0,0 +1,87 @@
+[aborting.any.html]
+  [WritableStreamDefaultController.signal]
```
Yes, I have added an item at https://github.com/servo/servo/issues/34676

jdm commented:
I have an in-progress review. I'm hoping to be finished on Monday.

gterzian committed:
add comments for steps in peek queue value

Signed-off-by: gterzian <2792687+gterzian@users.noreply.github.com>

jdm commented on components/script/dom/writablestreamdefaultwriter.rs:
```diff
@@ -0,0 +1,490 @@
+        error: SafeHandleValue,
+        can_gc: CanGc,
+    ) {
+        let mut ready_promise = self.ready_promise.borrow_mut();
+        // If writer.[[readyPromise]].[[PromiseState]] is "pending", reject writer.[[readyPromise]] with error.
+        if ready_promise.is_pending() {
+            ready_promise.reject_native(&error);
+        } else {
+            // Otherwise, set writer.[[readyPromise]] to a promise rejected with error.
+            let promise = Promise::new(global, can_gc);
+            promise.reject_native(&error);
```
GC risk with active mutable borrow.

gterzian replied:
Done

jdm commented on components/script/dom/writablestreamdefaultwriter.rs:
```diff
@@ -0,0 +1,490 @@
+        error: SafeHandleValue,
+        can_gc: CanGc,
+    ) {
+        let mut closed_promise = self.closed_promise.borrow_mut();
+        // If writer.[[closedPromise]].[[PromiseState]] is "pending", reject writer.[[closedPromise]] with error.
+        if closed_promise.is_pending() {
+            closed_promise.reject_native(&error);
+        } else {
+            // Otherwise, set writer.[[closedPromise]] to a promise rejected with error.
+            let promise = Promise::new(global, can_gc);
+            promise.reject_native(&error);
```
GC risk with active mutable borrow.

gterzian replied:
Done, also in a few other place in the file.

jdm commented on components/script/dom/writablestreamdefaultwriter.rs:
```diff
@@ -0,0 +1,490 @@
+        };
+
+        // Return ! WritableStreamAbort(stream, reason).
+        rooted!(in(*cx) let mut reason_clone = UndefinedValue());
+        reason_clone.set(reason.get());
```
This should be unnecessary with my suggestion for abort.

gterzian replied:
Done

jdm commented on components/script/dom/writablestreamdefaultwriter.rs:
```diff
@@ -0,0 +1,490 @@
+        };
+
+        // Assert: stream.[[writer]] is writer.
+        assert!(stream.get_writer().map_or(false, |writer| &*writer == self));
```
```suggestion
        assert!(stream.get_writer().is_some_and(|writer| &*writer == self));
```

gterzian replied:
Done

jdm reviewed (approved):
Really nice work! It's a big PR, but it was written in a way that made the code easy to follow and match against the specification.

gterzian committed:
fix doc link for the abort algorihtm fulfillment handler

Signed-off-by: gterzian <2792687+gterzian@users.noreply.github.com>

gterzian commented on components/script/dom/writablestream.rs:
```diff
@@ -0,0 +1,949 @@
+
+impl Callback for AbortAlgorithmRejectionHandler {
+    fn callback(&self, cx: SafeJSContext, v: SafeHandleValue, _realm: InRealm, _can_gc: CanGc) {
+        // Reject abortRequest’s promise with undefined.
```
This seems incorrect, see 13.1 of https://streams.spec.whatwg.org/#writable-stream-finish-erroring

gterzian replied:
Done

gterzian committed:
fix step doc and variable name in abort algo rejection handler

Signed-off-by: gterzian <2792687+gterzian@users.noreply.github.com>

gterzian commented on components/script/dom/writablestream.rs:
```diff
@@ -0,0 +1,949 @@
+}
+
+impl Callback for AbortAlgorithmRejectionHandler {
+    fn callback(&self, cx: SafeJSContext, v: SafeHandleValue, _realm: InRealm, _can_gc: CanGc) {
```
DOne

gterzian committed:
Add must_root to pending abort request

Co-authored-by: Josh Matthews <josh@joshmatthews.net>
Signed-off-by: Gregory Terzian <2792687+gterzian@users.noreply.github.com>

gterzian committed:
limit visibility to crate for has_operations_marked_inflight

Co-authored-by: Josh Matthews <josh@joshmatthews.net>
Signed-off-by: Gregory Terzian <2792687+gterzian@users.noreply.github.com>

gterzian committed:
limit visibility to crate for get_stored_error

Co-authored-by: Josh Matthews <josh@joshmatthews.net>
Signed-off-by: Gregory Terzian <2792687+gterzian@users.noreply.github.com>

gterzian committed:
remove potention re-borrow risk in reject loop on write requests in finish_erroring

Signed-off-by: gterzian <2792687+gterzian@users.noreply.github.com>

gterzian committed:
remove potential re-borrow risk when taking pending abort request in finish_erroring

Signed-off-by: gterzian <2792687+gterzian@users.noreply.github.com>

gterzian commented on components/script/dom/writablestream.rs:
```diff
@@ -0,0 +1,949 @@
+        self.get_stored_error(stored_error.handle_mut());
+
+        // For each writeRequest of stream.[[writeRequests]]:
+        for request in self.write_requests.borrow_mut().drain(..) {
+            // Reject writeRequest with storedError.
+            request.reject(cx, stored_error.handle());
```
DOne

gterzian committed:
remove potential re-borrow risk when taking close request in reject_close_and_closed_promise_if_needed

Signed-off-by: gterzian <2792687+gterzian@users.noreply.github.com>

gterzian committed:
remove re-borrow risks in finish_in_flight_close

Signed-off-by: gterzian <2792687+gterzian@users.noreply.github.com>

gterzian committed:
remove re-borrow risk on in_flight_close_request in finish_in_flight_close_with_error

Signed-off-by: gterzian <2792687+gterzian@users.noreply.github.com>

gterzian committed:
remove unnecessary clone of of reason in abort

Signed-off-by: gterzian <2792687+gterzian@users.noreply.github.com>

gterzian commented on components/script/dom/writablestream.rs:
```diff
@@ -0,0 +1,949 @@
+        // If wasAlreadyErroring is false,
+        if !was_already_erroring {
+            rooted!(in(*cx) let mut reason_clone = UndefinedValue());
+            reason_clone.set(reason.get());
```
Done

gterzian commented on components/script/dom/writablestream.rs:
```diff
@@ -0,0 +1,949 @@
+        // If stream.[[pendingAbortRequest]] is not undefined,
+        if let Some(pending_abort_request) = self.pending_abort_request.borrow_mut().take() {
+            // Reject stream.[[pendingAbortRequest]]'s promise with error.
+            pending_abort_request.promise.reject_native(&error);
```
Done

gterzian commented on components/script/dom/writablestream.rs:
```diff
@@ -0,0 +1,949 @@
+            // If stream.[[pendingAbortRequest]] is not undefined,
+            if let Some(pending_abort_request) = self.pending_abort_request.borrow_mut().take() {
+                // Resolve stream.[[pendingAbortRequest]]'s promise with undefined.
+                pending_abort_request.promise.resolve_native(&());
```
Done

gterzian commented on components/script/dom/writablestream.rs:
```diff
@@ -0,0 +1,949 @@
+        in_flight_close_request
+            .as_ref()
+            .unwrap()
+            .resolve_native(&());
```
Done

gterzian committed:
fix condition on backpressure and a writable state in close

Signed-off-by: gterzian <2792687+gterzian@users.noreply.github.com>

gterzian commented on components/script/dom/writablestream.rs:
```diff
@@ -0,0 +1,949 @@
+        if let Some(writer) = self.writer.get() {
+            // and stream.[[backpressure]] is true,
+            // and state is "writable",
+            if self.get_backpressure() || self.is_writable() {
```
DOne

gterzian committed:
limit visibility to crate for update_backpressure

Co-authored-by: Josh Matthews <josh@joshmatthews.net>
Signed-off-by: Gregory Terzian <2792687+gterzian@users.noreply.github.com>

gterzian commented on components/script/dom/writablestream.rs:
```diff
@@ -0,0 +1,949 @@
+            UnderlyingSink::empty()
+        };
+
+        if !underlying_sink_dict.type_.handle().is_undefined() {
```
It seems so: 

```
no method named `is_undefined` found for struct `trace::RootedTraceableBox<mozjs_sys::jsgc::Heap<js::jsapi::Value>>` in the current scope
```

gterzian committed:
remove mutability of reason in abort workflow

Signed-off-by: gterzian <2792687+gterzian@users.noreply.github.com>

gterzian commented on components/script/dom/writablestream.rs:
```diff
@@ -0,0 +1,949 @@
+        }
+
+        rooted!(in(*cx) let mut reason_clone = UndefinedValue());
+        reason_clone.set(reason.get());
```
Done

gterzian committed:
remove unnecessary use of ignore_malloc_size_of around Dom in controller

Signed-off-by: gterzian <2792687+gterzian@users.noreply.github.com>

gterzian commented on components/script/dom/writablestreamdefaultcontroller.rs:
```diff
@@ -0,0 +1,834 @@
+#[derive(Clone, JSTraceable, MallocSizeOf)]
+#[cfg_attr(crown, crown::unrooted_must_root_lint::must_root)]
+struct StartAlgorithmFulfillmentHandler {
+    #[ignore_malloc_size_of = "Trusted are hard"]
```
Done

gterzian commented on components/script/dom/writablestreamdefaultcontroller.rs:
```diff
@@ -0,0 +1,834 @@
+#[derive(Clone, JSTraceable, MallocSizeOf)]
+#[cfg_attr(crown, crown::unrooted_must_root_lint::must_root)]
+struct StartAlgorithmRejectionHandler {
+    #[ignore_malloc_size_of = "Trusted are hard"]
```
Done

gterzian commented on components/script/dom/writablestreamdefaultcontroller.rs:
```diff
@@ -0,0 +1,834 @@
+#[derive(Clone, JSTraceable, MallocSizeOf)]
+#[cfg_attr(crown, crown::unrooted_must_root_lint::must_root)]
+struct WriteAlgorithmFulfillmentHandler {
+    #[ignore_malloc_size_of = "Trusted are hard"]
```
Done

gterzian commented on components/script/dom/writablestreamdefaultcontroller.rs:
```diff
@@ -0,0 +1,834 @@
+#[derive(Clone, JSTraceable, MallocSizeOf)]
+#[cfg_attr(crown, crown::unrooted_must_root_lint::must_root)]
+struct WriteAlgorithmRejectionHandler {
+    #[ignore_malloc_size_of = "Trusted are hard"]
```
Done

gterzian committed:
fix ignore malloc size of comment for strategy size

Signed-off-by: gterzian <2792687+gterzian@users.noreply.github.com>

gterzian commented on components/script/dom/writablestreamdefaultcontroller.rs:
```diff
@@ -0,0 +1,834 @@
+    strategy_hwm: f64,
+
+    /// <https://streams.spec.whatwg.org/#writablestreamdefaultcontroller-strategysizealgorithm>
+    #[ignore_malloc_size_of = "mozjs"]
```
Done

gterzian committed:
reduce visibility of public methods to crate in controller

Signed-off-by: gterzian <2792687+gterzian@users.noreply.github.com>

gterzian committed:
remove use of JS_GetPendingException in controller get_chunk_size

Signed-off-by: gterzian <2792687+gterzian@users.noreply.github.com>

gterzian committed:
return early on error in write

Signed-off-by: gterzian <2792687+gterzian@users.noreply.github.com>

gterzian committed:
use is_some_and in assertion that stream.witer is writer in release

Signed-off-by: gterzian <2792687+gterzian@users.noreply.github.com>

gterzian commented:
@jdm @Taym95 Thanks for the review, I have addressed all comments in separate commits. I have also made additional fixes of the same kind in https://github.com/servo/servo/pull/34844/commits/8ab68e88579087c59a6ee91cdb9f024367cd3b5f

gterzian committed:
root pending abort request

uses ai

Signed-off-by: gterzian <2792687+gterzian@users.noreply.github.com>

gterzian committed:
fix mutable re-borrow risk in writer

Signed-off-by: gterzian <2792687+gterzian@users.noreply.github.com>
